ocaml/typing/solver.ml

```diff
@@ -0,0 +1,724 @@
+        (* TODO: mabye create a intermediate variable? *)
+        Result.bind (submode_cc a b) (fun () ->
+            Result.bind
+              (find_error (fun mv -> submode_mvc mv b) mvs)
+              (fun () ->
+                Result.bind
+                  (find_error (fun mu -> submode_cmv a mu) mus)
+                  (fun () ->
+                    find_error
+                      (fun mu -> find_error (fun mv -> submode_mvmv mv mu) mvs)
+                      mus)))
+    in
+    match r with
+    | Ok () -> Ok log
+    | Error e ->
+      (* we mutated some states and found conflict;
+         need to revert those mutation to keep the state consistent.
+         A nice by-effect is that this function doesn't mutate state in failure
+      *)
+      Option.iter (fun log -> undo_changes !log) log;
+      Error e
+
+  let submode obj a b =
+    match submode_try obj a b with
+    | Ok log ->
+      Option.iter !append_changes log;
+      Ok ()
+    | Error _ as e -> e
+
+  let zap_to_ceil_morphvar obj mv =
+    assert (submode obj (Amode (mupper obj mv)) (Amodevar mv) |> Result.is_ok);
+    mupper obj mv
+
+  let zap_to_ceil : type a l. a C.obj -> (a, l * allowed) mode -> a =
+   fun obj -> function
+    | Amode m -> m
+    | Amodevar mv -> zap_to_ceil_morphvar obj mv
+    | Amodemeet (a, mvs) ->
+      List.fold_left
+        (fun acc mv -> C.meet obj acc (zap_to_ceil_morphvar obj mv))
+        a mvs
+
+  let join (type a r) obj l =
+    let rec loop :
+        a ->
+        (a, allowed * disallowed) morphvar list ->
+        (a, allowed * r) mode list ->
+        (a, allowed * disallowed) mode =
+     fun a mvs ->
+      if C.le obj (C.max obj) a
+      then fun _ -> Amode (C.max obj)
+      else
+        function
+        | [] -> Amodejoin (a, mvs)
+        | mv :: xs -> (
+          match disallow_right mv with
+          | Amode b -> loop (C.join obj a b) mvs xs
+          | Amodevar mv -> loop a (mv :: mvs) xs
+          | Amodejoin (b, mvs') -> loop (C.join obj a b) (mvs' @ mvs) xs)
+    in
+    loop (C.min obj) [] l
+
+  let meet (type a l) obj l =
+    let rec loop :
+        a ->
+        (a, disallowed * allowed) morphvar list ->
+        (a, l * allowed) mode list ->
+        (a, disallowed * allowed) mode =
+     fun a mvs ->
+      if C.le obj a (C.min obj)
+      then fun _ -> Amode (C.min obj)
+      else
+        function
+        | [] -> Amodemeet (a, mvs)
+        | mv :: xs -> (
+          match disallow_left mv with
+          | Amode b -> loop (C.meet obj a b) mvs xs
+          | Amodevar mv -> loop a (mv :: mvs) xs
+          | Amodemeet (b, mvs') -> loop (C.meet obj a b) (mvs' @ mvs) xs)
+    in
+    loop (C.max obj) [] l
+
+  let zap_to_floor_morphvar ~commit obj mv =
+    let log, lower = constrain_mlower_try obj mv in
+    if commit then !append_changes log else undo_changes !log;
+    lower
+
+  let zap_to_floor : type a r. a C.obj -> (a, allowed * r) mode -> a =
```

Comment:
I had thought that this function also mutates the `mode` to be its minimum value. But it doesn't do that, in the `Amodejoin` case. Maybe that's fine... but now I'm not sure about the name of the function again.

Reply:
I'm not sure about this - I think in the `Amodejoin` case it still mutates the `mode` to be minimum?

Reply:
I agree with you here. Not sure what I saw last time.